### PR TITLE
provider/datadog: Add datadog_user resource

### PR DESCRIPTION
**Pull request by bflad**
Adds support for managing Datadog users.

```
make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestAccDatadogUser'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/26 18:56:47 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestAccDatadogUser -timeout 120m
=== RUN   TestAccDatadogUser_Required
--- PASS: TestAccDatadogUser_Required (1.82s)
=== RUN   TestAccDatadogUser_Email
--- PASS: TestAccDatadogUser_Email (1.78s)
=== RUN   TestAccDatadogUser_IsAdmin
--- PASS: TestAccDatadogUser_IsAdmin (2.00s)
=== RUN   TestAccDatadogUser_NameUpdated
--- PASS: TestAccDatadogUser_NameUpdated (4.14s)
=== RUN   TestAccDatadogUser_Role
--- SKIP: TestAccDatadogUser_Role (0.00s)
	resource_datadog_user_test.go:116: Datadog API does not support updating User role.
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/datadog	9.751s
```

/cc @ojongerius

### CODE REVIEW

**Comment by ojongerius** [comment]
Awesome!

Afraid `TestAccDatadogUser_Required` on 1383ea96f817e6660c92fe1d97e6e90428deedbb fails for me:

```
▶ make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestDatadogUser'       
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/27 13:27:14 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestDatadogUser -timeout 120m
=== RUN   TestDatadogUser_import
--- FAIL: TestDatadogUser_import (0.00s)
        testing.go:265: Step 0 error: Configuration is invalid.
                
                Warnings: []string(nil)
                
                Errors: []string{"datadog_user.foo: \"email\": required field is not set"}
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/builtin/providers/datadog        0.015s
make: *** [testacc] Error 1
```

See that acceptance tests are passing indeed:

```
github.com/hashicorp/terraform  datadog_user_resource ✔                                                                                                                                                                                                                                 8h40m  
▶ make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestAccDatadogUser'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/27 13:35:07 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestAccDatadogUser -timeout 120m
=== RUN   TestAccDatadogUser_Required
2017/02/27 13:35:11 [WARN] Invalid log level: "1". Defaulting to level: TRACE. Valid levels are: [TRACE DEBUG INFO WARN ERROR]
--- PASS: TestAccDatadogUser_Required (8.55s)
=== RUN   TestAccDatadogUser_Email
--- PASS: TestAccDatadogUser_Email (7.03s)
=== RUN   TestAccDatadogUser_IsAdmin
--- PASS: TestAccDatadogUser_IsAdmin (7.42s)
=== RUN   TestAccDatadogUser_NameUpdated
--- PASS: TestAccDatadogUser_NameUpdated (12.10s)
=== RUN   TestAccDatadogUser_Role
--- SKIP: TestAccDatadogUser_Role (0.00s)
        resource_datadog_user_test.go:116: Datadog API does not support updating User role.
PASS
ok      github.com/hashicorp/terraform/builtin/providers/datadog        35.119s
```

File: builtin/providers/datadog/import_datadog_user_test.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+const testAccCheckDatadogUserConfigImported = `
+resource "datadog_user" "foo" {
+  handle = "test@example.com"
+  name   = "Test User"
```
Comment (ojongerius) [inline builtin/providers/datadog/import_datadog_user_test.go]: Do you have to add `email = foo@bar.org` to make this pass? Currently:

```
▶ make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestDatadogUser'            
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/27 13:38:18 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestDatadogUser -timeout 120m
=== RUN   TestDatadogUser_import
--- FAIL: TestDatadogUser_import (0.00s)
        testing.go:265: Step 0 error: Configuration is invalid.
                
                Warnings: []string(nil)
                
                Errors: []string{"datadog_user.foo: \"email\": required field is not set"}
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/builtin/providers/datadog        0.015s
make: *** [testacc] Error 1

```

File: builtin/providers/datadog/resource_datadog_user.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	var u datadog.User
+	u.SetDisabled(d.Get("disabled").(bool))
+	u.SetEmail(d.Get("email").(string))
+	u.SetHandle(d.Get("handle").(string))
+	u.SetIsAdmin(d.Get("is_admin").(bool))
+	u.SetName(d.Get("name").(string))
+	u.SetRole(d.Get("role").(string))
+
+	// Datadog does not actually delete users, so CreateUser might return a 409.
+	// We ignore that case and proceed, likely re-enabling the user.
```
Comment (ojongerius) [inline builtin/providers/datadog/resource_datadog_user.go]: Should we inform users that we are recreating a user?

File: builtin/providers/datadog/resource_datadog_user.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	}
+
+	return resourceDatadogUserRead(d, meta)
+}
+
+func resourceDatadogUserDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*datadog.Client)
+
+	// Datadog does not actually delete users. If user is already disabled,
+	// the Go API retries for awhile (1+ minute) then errors with a 400.
```
Comment (ojongerius) [inline builtin/providers/datadog/resource_datadog_user.go]: Is retrying current behaviour? I know this was the case before https://github.com/zorkian/go-datadog-api/pull/95 was merged. If it's a 400 we should not retry, see https://github.com/zorkian/go-datadog-api/blob/master/request.go#L140

File: builtin/providers/datadog/resource_datadog_user_test.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+						"datadog_user.foo", "handle", "test@example.com"),
+					resource.TestCheckResourceAttr(
+						"datadog_user.foo", "name", "Test User"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccDatadogUser_Email(t *testing.T) {
```
Comment (ojongerius) [inline builtin/providers/datadog/resource_datadog_user_test.go]: What is this test adding to `TestAccDatadogUser_Required`?

File: builtin/providers/datadog/resource_datadog_user_test.go [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+						"datadog_user.foo", "handle", "test@example.com"),
+					resource.TestCheckResourceAttr(
+						"datadog_user.foo", "name", "Test User"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccDatadogUser_IsAdmin(t *testing.T) {
```
Comment (ojongerius) [inline builtin/providers/datadog/resource_datadog_user_test.go]: Same as previous remark, I'd consider collapsing `TestAccDatadogUser_Required`, `TestAccDatadogUser_Email` and `TestAccDatadogUser_IsAdmin`.

File: builtin/providers/datadog/resource_datadog_user_test.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+						"datadog_user.foo", "name", "Test User"),
+					resource.TestCheckResourceAttr(
+						"datadog_user.foo", "is_admin", "true"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccDatadogUser_NameUpdated(t *testing.T) {
```
Comment (ojongerius) [inline builtin/providers/datadog/resource_datadog_user_test.go]: Might be worthwhile updating `is_admin` too.

File: builtin/providers/datadog/resource_datadog_user_test.go [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+					resource.TestCheckResourceAttr(
+						"datadog_user.foo", "name", "Updated User"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccDatadogUser_Role(t *testing.T) {
+	t.Skip("Datadog API does not support updating User role.")
```
Comment (ojongerius) [inline builtin/providers/datadog/resource_datadog_user_test.go]: Not sure if this code is worth having if we can't use it. Was the the rationale to ad it?

File: website/source/docs/providers/datadog/r/user.html.markdown [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `disabled` - (Optional) Whether the user is disabled
+* `email` - (Required) Email address for user
+* `handle` - (Required) Datadog ID for user (generally same as email)
```
Comment (ojongerius) [inline website/source/docs/providers/datadog/r/user.html.markdown]: According to DD API docs: `The user handle, must be a valid email.` Might be worth adding as I'm assuming they will return 400 if it does not look like an email address, and users might be confused if it fails.

File: builtin/providers/datadog/resource_datadog_user.go (hunk 2 above)
Comment (bflad) [inline builtin/providers/datadog/resource_datadog_user.go]: Is something like this the appropriate way to do that?
```
log.Printf("[INFO] Updating existing Datadog user %q", u.Handle)
```

**Comment by bflad** [comment]
I've pushed up a new commit with the requested changes. Please let me know if there are other things that need adjusting. Thanks!

```
make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestAccDatadogUser'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/26 20:17:53 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestAccDatadogUser -timeout 120m
=== RUN   TestAccDatadogUser_Updated
--- PASS: TestAccDatadogUser_Updated (2.96s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/datadog	2.983s


make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestDatadogUser'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/26 20:18:27 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestDatadogUser -timeout 120m
=== RUN   TestDatadogUser_import
--- PASS: TestDatadogUser_import (2.11s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/datadog	2.130s
```

File: website/source/docs/providers/datadog/r/user.html.markdown [hunk 9]
```diff
@@ -27,7 +27,7 @@ The following arguments are supported:
 
 * `disabled` - (Optional) Whether the user is disabled
 * `email` - (Required) Email address for user
-* `handle` - (Required) Datadog ID for user (generally same as email)
+* `handle` - (Required) The user handle, must be a valid email.
```
Comment (ojongerius) [inline website/source/docs/providers/datadog/r/user.html.markdown]: Nit; "must be a valid email address". After all we are interested in an *address*, not some email random email 😉

**Comment by ojongerius** [comment]
All tests pass, 🏆!

```
▶ make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestDatadogUser'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/27 15:34:51 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestDatadogUser -timeout 120m
=== RUN   TestDatadogUser_import
--- PASS: TestDatadogUser_import (10.77s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/datadog        10.784s

github.com/hashicorp/terraform  0-8-stable-datadog_user_resource ✔                                                                                                                                                                                                         3h13m  
▶ make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestAccDatadogUser'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/27 15:39:38 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestAccDatadogUser -timeout 120m
=== RUN   TestAccDatadogUser_Updated
--- PASS: TestAccDatadogUser_Updated (15.07s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/datadog        15.082s
```

**Pull request merged**